File: examples/httpServer/output.rb
require 'sinatra'
require 'pg'

# Initiate HTTP server on a specified port
set :port, 8080

# Establish a connection to the PostgreSQL database
$db = PG.connect("postgres://admin:pass@localhost/test?sslmode=disable")

# Handle errors and output the logs
set :show_exceptions, :after_handler

error do
  'Sorry there was a nasty error - ' + env['sinatra.error'].message
end

# Create a new user and submit it to the database
post '/api/create' do
  username = params[:username]
  password = params[:password]

  # Ensure username and password are valid
  halt 400, 'Invalid parameters' if username.nil? || password.nil?

  # Insert new user record into PostgreSQL
  $db.exec_params("INSERT INTO users (username, password) VALUES ($1, $2)", [username, password])

  # Return 200 OK if created or error if not
  status 200
end

# Update an existing user's password based on the username
put '/api/password_change' do
  username = params[:username]
  newpassword = params[:newpassword]

  halt 400, 'Invalid parameters' if username.nil? || newpassword.nil?

  $db.exec_params("UPDATE users SET password = $1 WHERE username = $2", [newpassword, username])

  status 200
end

# Return 200 OK if user found, and 404 if it isn't
get '/api/retrieve' do
  username = params[:username]

  halt 400, 'Invalid parameters' if username.nil?

  res = $db.exec_params("SELECT * FROM users WHERE username = $1", [username])

  halt 404, 'User not found' if res.ntuples.zero?

  status 200
end

# Remove a user record from the database based on the username
delete '/api/remove' do
  username = params[:username]

  halt 400, 'Invalid parameters' if username.nil?

  $db.exec_params("DELETE FROM users WHERE username = $1", [username])

  status 200
end

File: examples/httpServer/output.php
<?php

// Start HTTP Server
$server = new swoole_http_server("127.0.0.1", 8080);

// Establish a connection to the PostgreSQL database
$db = new PDO('pgsql:host=localhost;dbname=test', 'admin', 'pass');

$server->on('request', function ($request, $response) use ($db) {
    // Validate request
    if (empty($request->get) || empty($request->post)) {
        $response->status(400);
        $response->end('Invalid request');
        return;
    }

    // Handle routes
    switch ($request->server['request_uri']) {
        case '/api/create':
            // Ensure username and password are valid
            $username = $request->post['username'];
            $password = $request->post['password'];

            if (empty($username) || empty($password)) {
                $response->status(400);
                $response->end('Invalid parameters');
                return;
            }

            // Insert new user record into PostgreSQL
            $stmt = $db->prepare('INSERT INTO users (username, password) VALUES (?, ?)');
            $stmt->execute([$username, $password]);

            // Return 200 OK if created or error if not
            $response->status(200);
            $response->end('User created');
            break;
        case '/api/password_change':
            // Update an existing user's password based on the username
            $username = $request->post['username'];
            $newpassword = $request->post['newpassword'];

            $stmt = $db->prepare('UPDATE users SET password = ? WHERE username = ?');
            $stmt->execute([$newpassword, $username]);

            $response->status(200);
            $response->end('Password updated');
            break;
        case '/api/retrieve':
            // Return 200 OK if user found, and 404 if it isn't
            $username = $request->get['username'];

            $stmt = $db->prepare('SELECT * FROM users WHERE username = ?');
            $stmt->execute([$username]);

            if ($stmt->rowCount() > 0) {
                $response->status(200);
                $response->end('User found');
            } else {
                $response->status(404);
                $response->end('User not found');
            }
            break;
        case '/api/remove':
            // Remove a user record from the database based on the username
            $username = $request->get['username'];

            $stmt = $db->prepare('DELETE FROM users WHERE username = ?');
            $stmt->execute([$username]);

            $response->status(200);
            $response->end('User removed');
            break;
    }
});

$server->start();

?>
